File: app/src/main/java/torille/fi/lurkforreddit/comments/CustomUrlSpan.java
package torille.fi.lurkforreddit.comments;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Parcel;
import android.text.style.URLSpan;
import android.view.View;

import org.parceler.Parcels;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import timber.log.Timber;
import torille.fi.lurkforreddit.data.models.Post;
import torille.fi.lurkforreddit.data.models.PostDetails;
import torille.fi.lurkforreddit.media.FullscreenActivity;
import torille.fi.lurkforreddit.subreddit.SubredditActivity;
import torille.fi.lurkforreddit.utils.MediaHelper;

/**
 * Custom {@link URLSpan} to modify how to open clicked links in text
 */

class CustomUrlSpan extends URLSpan {

    CustomUrlSpan(String url) {
        super(url);
    }

    @Override
    public void onClick(View widget) {
        String url = getURL();
        Uri uri = Uri.parse(url);
        String domain = uri.getHost();
        Timber.d("got url " + url);

        Intent intent;
        Context context = widget.getContext();
        if (MediaHelper.isContentMedia(url) || MediaHelper.checkDomainForMedia(domain)) {
            Post post = new Post("t5", new PostDetails());
            post.getPostDetails().setUrl(url);
            post.getPostDetails().setDomain(domain);

            intent = new Intent(context, FullscreenActivity.class);
            intent.putExtra(FullscreenActivity.EXTRA_POST, Parcels.wrap(post));

            context.startActivity(intent);
        } else if (checkForReddit(url)) {

            intent = new Intent(context, SubredditActivity.class);
            intent.putExtra(SubredditActivity.EXTRA_SUBREDDITNAME, url);

            Timber.d("Going to checkout subreddit " + url);

            context.startActivity(intent);
        } else {
            super.onClick(widget);
        }


    }

    private boolean checkForReddit(String redditUrl) {
        Timber.d("Checking url " + redditUrl);
        Pattern p = Pattern.compile("(/r/.*)");
        Matcher m = p.matcher(redditUrl);
        return m.matches();
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
    }

    protected CustomUrlSpan(Parcel in) {
        super(in);
    }

    public static final Creator<CustomUrlSpan> CREATOR = new Creator<CustomUrlSpan>() {
        @Override
        public CustomUrlSpan createFromParcel(Parcel source) {
            return new CustomUrlSpan(source);
        }

        @Override
        public CustomUrlSpan[] newArray(int size) {
            return new CustomUrlSpan[size];
        }
    };
}

File: app/src/main/java/torille/fi/lurkforreddit/comments/CustomUrlSpan.kt
package torille.fi.lurkforreddit.comments

import android.content.Intent
import android.graphics.Canvas
import android.graphics.Paint
import android.net.Uri
import android.text.Layout
import android.text.style.LeadingMarginSpan
import android.text.style.URLSpan
import android.view.View
import timber.log.Timber
import torille.fi.lurkforreddit.data.models.view.Post
import torille.fi.lurkforreddit.media.FullscreenActivity
import torille.fi.lurkforreddit.utils.AppLinkActivity
import torille.fi.lurkforreddit.utils.MediaHelper
import java.util.regex.Pattern


/**
 * Custom [URLSpan] to modify how to open clicked links in text
 */

internal class CustomUrlSpan(url: String, val color: Int) : URLSpan(url), LeadingMarginSpan {

    override fun drawLeadingMargin(canvas: Canvas?, paint: Paint?, x: Int, dir: Int, top: Int, baseline: Int, bottom: Int, p7: CharSequence?, p8: Int, p9: Int, p10: Boolean, p11: Layout?) {
        paint?.style = Paint.Style.FILL
        paint?.color = color
        canvas?.drawRect(x.toFloat(), x.toFloat(), x.toFloat().plus(dir).plus(1), bottom.toFloat(), paint!!)
    }


    override fun getLeadingMargin(p0: Boolean): Int {
        return 5
    }

    override fun onClick(widget: View) {
        val url = url
        val uri = Uri.parse(url)
        val domain = uri.host
        Timber.d("Got url $url")
        val intent: Intent
        val context = widget.context
        when {
            MediaHelper.isContentMedia(url, domain) -> {
                intent = Intent(context, FullscreenActivity::class.java)
                intent.putExtra(FullscreenActivity.EXTRA_POST, Post(url = url))
                intent.putExtra(FullscreenActivity.EXTRA_URL, url)
                context.startActivity(intent)
            }
            checkForReddit(url) -> {
                Timber.d("Going to checkout subreddit $url")
                intent = Intent(context, AppLinkActivity::class.java)
                intent.data = Uri.parse(url)
                context.startActivity(intent)
            }
            else -> super.onClick(widget)
        }

    }

    private fun checkForReddit(redditUrl: String): Boolean {
        Timber.d("Checking url $redditUrl")
        val p = Pattern.compile("(/r/.*)")
        val m = p.matcher(redditUrl)
        return m.matches()
    }
}
